crates/vm/src/system/memory/tree/mod.rs

```diff
@@ -66,22 +66,45 @@ impl<const CHUNK: usize, F: PrimeField32> MemoryNode<CHUNK, F> {
     }
 
     fn from_memory(
-        memory: &BTreeMap<u64, [F; CHUNK]>,
+        memory: &Vec<(u64, [F; CHUNK])>,
+        lookup_range: Range<usize>,
         height: usize,
         from: u64,
         hasher: &impl Hasher<CHUNK, F>,
     ) -> MemoryNode<CHUNK, F> {
-        let mut range = memory.range(from..from + (1 << height));
         if height == 0 {
-            let values = *memory.get(&from).unwrap_or(&[F::ZERO; CHUNK]);
-            MemoryNode::new_leaf(hasher.hash(&values))
-        } else if range.next().is_none() {
+            if lookup_range.is_empty() {
+                MemoryNode::new_leaf(hasher.hash(&[F::ZERO; CHUNK]))
+            } else {
+                debug_assert_eq!(memory[lookup_range.start].0, from);
+                debug_assert_eq!(lookup_range.end - lookup_range.start, 1);
+                MemoryNode::new_leaf(hasher.hash(&memory[lookup_range.start].1))
+            }
+        } else if lookup_range.is_empty() {
             let leaf_value = hasher.hash(&[F::ZERO; CHUNK]);
             MemoryNode::construct_uniform(height, leaf_value, hasher)
         } else {
             let midpoint = from + (1 << (height - 1));
-            let left = Self::from_memory(memory, height - 1, from, hasher);
-            let right = Self::from_memory(memory, height - 1, midpoint, hasher);
+            let mid = {
+                let mut left = lookup_range.start;
+                let mut right = lookup_range.end;
+                if memory[left].0 >= midpoint {
+                    left
+                } else {
+                    while left + 1 < right {
+                        let mid = left + (right - left) / 2;
+                        if memory[mid].0 < midpoint {
+                            left = mid;
+                        } else {
+                            right = mid;
+                        }
+                    }
+                    right
+                }
+            };
+            let left = Self::from_memory(memory, lookup_range.start..mid, height - 1, from, hasher);
```

Comment:
I think in principle you could do left, right in different threads

Reply:
https://github.com/openvm-org/openvm/pull/1341